### ISSUE TITLE
Resources fail to load in AssemblyLoadContext

### ISSUE THREAD

**Issue by jaredpar** [dotnet/runtime#2022]
To reproduce this issue:

- Download and unzip [console.zip](https://github.com/dotnet/runtime/files/4099313/console.zip)
- Open `Program.cs` and change line 45 to have the full path to where you unzipped the content `LoadFromAssemblyPath(@"p:\temp\console\Microsoft.VisualBasic.dll");`
- Execute `dotnet run` 

The expected output here is 

> Public member 'G' on type 'A' not found.

The actual output though is 

> Value cannot be null. (Parameter 'format')

The code here is a simplified example of Visual Basic late binding (essentially `dynamic`). The expectation is the late binding fails here but with an actionable error message. What happens in this case though is that the `Microsoft.VisualBasic.dll` is unable to load embedded resources and as a result ends up passing `null` to `string.Format` and generating the above exception.

One key thing to note is this repro is using the desktop `Microsoft.VisualBasic.dll`, not the .NET Core version. At the same time this seems like a fairly straight forward resource load and it's surprising that this does not work.

The context for this issue is the pattern comes up a bit in the Visual Basic Compiler test suites. The tests were initially written for desktop and when we ported them to run on Core we kept the desktop reference to Microsoft.VisualBasic (it worked hence why update). In attempting to move to netcoreapp3.1 though this behavior suddenly started changing and I'm attempting to understand why before we go through the process of re-engineering these tests.

**Comment by jeffschwMSFT** [comment]
cc @elinor-fung @vitek-karas

**Comment by vitek-karas** [comment]
cc @sdmaclea 
@jaredpar Do you know if the change in behavior happened in 3.0->3.1 or sometimes in 3.*?

We did change resource loading code a little bit, but it was meant to work as before.

**Comment by jaredpar** [comment]
@vitek-karas 

> Do you know if the change in behavior happened in 3.0->3.1 or sometimes in 3.*?

Not off hand. This particular set of projects never tested against `netcoreapp3.0`. I can try and repro against `netcoreapp3.0` and see if it happens.

**Comment by elinor-fung** [comment]
I just tried on 3.0 and saw the same `Value cannot be null. (Parameter 'format')` error.

The runtime throws:
```
System.Resources.MissingManifestResourceException: 'Could not find the resource "System.SR.resources" among the resources "FxResources.Microsoft.VisualBasic.SR.resources" embedded in the assembly "Microsoft.VisualBasic", nor among the resources in any satellite assemblies for the specified culture. Perhaps the resources were embedded with an incorrect name.'`
at
System.Private.CoreLib.dll!System.Resources.ManifestBasedResourceGroveler.HandleResourceStreamMissing(string fileName)
 System.Private.CoreLib.dll!System.Resources.ManifestBasedResourceGroveler.GrovelForResourceSet(System.Globalization.CultureInfo culture, System.Collections.Generic.Dictionary<string, System.Resources.ResourceSet> localResourceSets, bool tryParents, bool createIfNotExists)
 System.Private.CoreLib.dll!System.Resources.ResourceManager.InternalGetResourceSet(System.Globalization.CultureInfo culture, bool createIfNotExists, bool tryParents)
 System.Private.CoreLib.dll!System.Resources.ResourceManager.GetString(string name, System.Globalization.CultureInfo culture)
 System.Private.CoreLib.dll!System.Resources.ResourceManager.GetString(string name)
 Microsoft.VisualBasic.dll!System.SR.GetResourceString(string resourceKey, string defaultString)
 Microsoft.VisualBasic.dll!System.SR.MissingMember_MemberNotFoundOnType2.get()
```
Which presumably gets swallowed in Microsoft.VisualBasic.dll!System.SR.GetResourceString.

Is it 2.2 where this scenario was working with that Microsoft.VisualBasic.dll?

**Comment by sdmaclea** [comment]
@elinor-fung Looking at the code I wondered if the default context might be loading a different MS.VB.dll.  Maybe it is looking at the wrong assembly to get the resource???  It might be interesting to iterate through the Assemblies in the AppDomain just to be sure.

**Comment by sdmaclea** [comment]
It would also be nice to put  break point on line 45, to see that we actually load the Framework MS.VB.dll...

**Comment by elinor-fung** [comment]
It is correctly loading the MS.VB.dll that I downloaded from the repro and looking for the resource in that assembly. It looks like it using the wrong name `System.SR.resources` when `FxResources.Microsoft.VisualBasic.SR.resources` is the only one that actually exists.

When the `ResourceManager` used is created, it is initializied using the `System.SR` type, which what then gets used to determine resource name. Interestingly, looking at the DefinedTypes of the loaded MS.VB.dll under the debugger (or just in ildasm), it has a `FxResources.Microsoft.VisualBasic.SR` as well as `System.SR`.
```
-  resourceSource.Assembly  {Microsoft.VisualBasic, Version=10.0.2.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a}
-    DefinedTypes           {System.RuntimeType[78]}
     ...
+    [49]	{Name = "SR" FullName = "System.SR"}
+    [50]	{Name = "SR" FullName = "FxResources.Microsoft.VisualBasic.SR"}
```

If the `FxResources.Microsoft.VisualBasic.SR` had been the one received in the `ResourceManager` constructor, it would have looked for `FxResources.Microsoft.VisualBasic.SR.resources` (and found it). If that is actually the issue, I'm not sure what might have changed that would affect that though.

**Comment by danmosemsft** [comment]
Note that the message at least is much more helpful in the past, in that these days it lists the names of the streams that actually are in the assembly
`Could not find the resource "System.SR.resources" among the resources "FxResources.Microsoft.VisualBasic.SR.resources" embedded in the assembly...`

**Comment by jaredpar** [comment]
> Is it 2.2 where this scenario was working with that Microsoft.VisualBasic.dll?

This was working in `netcoreapp2.1` in the test harness that we use in the Roslyn Compiler test base. The repro I included with this bug is a simplification of that test harness. Was hoping it would be much easier to debug than having you all step through our compiler tests. 

At the same time though this bug still repros in `netcoreapp2.1` with the sample I included with this bug. Basically at this point I'm confused how it ever worked in the compiler test suite. 

I'm happy to sit down with someone and show them how to run the test in our compiler test bed if this simple repro isn't enough to figure out the problem.

**Comment by jaredpar** [comment]
Unfortunately i'm having trouble determining if this test passes in `netcoreapp3.0` or not. The VS test runner refuses to execute the test under `netcoreapp3.0` but is happy to do so for any other target framework. Also the CLI team can't provide me a way to force the runner to use `netcoreapp3.0` hence I'm stuck on that avenue as well.

**Comment by elinor-fung** [comment]
@jaredpar if you could point me at how to go through the relevant test where this is happening, I can take a look. From the simple repro, the failure seems expected across 3.* and older versions (MS.VB.dll always initializing the `ResourceManager` such that it looks for `System.SR.resources` instead of `FxResources.Microsoft.VisualBasic.SR.resources`)

**Comment by jaredpar** [comment]
@elinor-fung do you want to see the version where this succeeds on `netcoreapp2.1`?

> From the simple repro, the failure seems expected across 3.* and older versions 

Why though? This is simple resource loading code that worked on desktop for basically the existence of desktop (this DLL is as old as .NET). Why is core breaking the resource loading for this DLL? 

My worry here is whether this will negatively impact our general ability to run desktop DLLs on .NET Core. Essentially why is this resource pattern expected to fail?

**Comment by elinor-fung** [comment]
Yeah, I'd like to see it succeeding, because I don't understand how it would find the resource in any scenario from the simple repro. From what I could tell, it would always create a `ResourceManager` with the base name of a resource that doesn't exist - and I don't think that logic in the `ResourceManager` has really changed. What version of .NET is that DLL from?

When I tried the same scenario with full framework (4.8) using the MS.VB.dll from the framework install, the MS.VB.dll goes through a completely different path - creating a `ResourceManager` by explicitly passing in the resource base name, which does correspond to the embedded resource (named `Microsoft.VisualBasic.resources` in this case) actually in the assembly.

**Comment by danmosemsft** [comment]
Just an observer here but it seems this bug is going to @elinor-fung on the assumption that the resource manager is at fault. 

Isn't it more likely that in 2.1 it was embedded with the correct base name and in 3.1 it is not. It's hard to imagine how it could have been working before if the resource base name as also wrong then.

The choice of the name that resources are embedded with is controlled by properties project file and the managed targets and tasks.

**Comment by danmosemsft** [comment]
If I misunderstood, please disregard...

**Comment by jaredpar** [comment]
@danmosemsft 

> The choice of the name that resources are embedded with is controlled by properties project file and the managed targets and tasks.

This is a desktop DLL that we consume but do not build. Essentially I grabbed a copy out of the net472 GAC and referenced it in the project. Hence project properties shouldn't be coming into play here.

**Comment by elinor-fung** [comment]
@jaredpar and I looked at this together and it turned out that when the tests were succeeding on `netcoreapp2.1`, they were actually loading the .NET Core version of MS.VB.dll, not the desktop one.

**Comment by jaredpar** [comment]
> @jaredpar and I looked at this together and it turned out that when the tests were succeeding on netcoreapp2.1, they were actually loading the .NET Core version of MS.VB.dll, not the desktop one.

Indeed. The `AssemblyLoadContext` implementation in our tests use the TPA as part of it's decision on whether to load the desktop or .NET Core version of a given assembly. The TPA changed between `netcoreapp2.1` and `netcoreapp3.0` such that the `Microsoft.VisualBasic.dll` present in our output directory was added to the TPA list. Hence our test code began preferring that, a desktop version of Microsoft.VisualBasic, over the core ones.

So basically nothing changed here with respect to `AssemblyLoadContext`. I had mistakenly assumed we were always using the desktop version of Microsoft.VisualBasic on that test.
